### PR TITLE
Skill Trees: Additional 'Skill Tree' Implementation

### PR DESCRIPTION
**Pull request by VannStrasen**
**Following up on [our previous PR](https://github.com/uchicago-cs/chiventure/pull/697)**
Our most recent PR – the initial skill tree code creation (#697) – implemented a major part of our skill tree functionality. However, it did not fully implement all functions because doing so would be too much for one PR, so we split up the task across different sprints. 

**Specific changes since the last PR**
* A handful of functions inside the skilltree.c file were implemented.
* Some small changes were made to the return types of functions. 
* Header comments in skilltree.h were changed when this implementation phase discovered more return types than previously thought necessary.

**Notes to PR reviewers**
* While this looks like there is a lot of code added to dev, a large portion of this code should instead be reviewed in PR #697. (Refer to the specific changes for what was actually edited)
* These skills have yet to be tested, and that will be done by @steeltrapp and @michaelleestuy later in the sprint.

### CODE REVIEW

**Comment by VannStrasen** [comment]
Fixed the warnings generated from the merge into dev!

File: src/skilltrees/skilltree.c [hunk 1]
```diff
@@ -102,48 +102,199 @@ int skill_tree_free(skill_tree_t* tree) {
...
+        return NULL;
+    } else {
+        return acquired;
+    }
 }
 
 /* See skilltree.h */
 skill_t** skill_prereqs_missing(skill_tree_t* tree,
                                skill_inventory_t* inventory, sid_t sid,
                                int* nmissing) {
```
Comment (firatciftci) [inline src/skilltrees/skilltree.c:233]: To preserve the same style in variable naming convention, `nmissing` should be changed to something like `num_missing`.

File: src/skilltrees/skilltree.c [hunk 2]
```diff
@@ -102,48 +102,199 @@ int skill_tree_free(skill_tree_t* tree) {
...
-    /* TO DO */
-    return NULL;
+    assert(tree != NULL && inventory != NULL);
+
+    unsigned int nprereqs;
```
Comment (firatciftci) [inline src/skilltrees/skilltree.c:236]: To preserve the same style in variable naming convention, `nprereqs` should be changed to something like `num_prereqs`.

File: src/skilltrees/skilltree.c [hunk 3]
```diff
@@ -102,48 +102,199 @@ int skill_tree_free(skill_tree_t* tree) {
...
 }
 
 /* See skilltree.h */
 int inventory_skill_acquire(skill_tree_t* tree, skill_inventory_t* inventory,
                            skill_t* skill) {
-    /* TO DO */
-    return 0;
+    assert(tree != NULL && inventory != NULL && skill != NULL);
+
+    unsigned int nmissing;
```
Comment (firatciftci) [inline src/skilltrees/skilltree.c:285]: To preserve the same style in variable naming convention, `nmissing` should be changed to something like `num_missing`.

File: src/skilltrees/skilltree.c (hunk 1 above)
Comment (VannStrasen) [reply src/skilltrees/skilltree.c]: Fixed in most recent commit.

File: src/skilltrees/skilltree.c (hunk 2 above)
Comment (VannStrasen) [reply src/skilltrees/skilltree.c]: Fixed in most recent commit.

File: src/skilltrees/skilltree.c (hunk 3 above)
Comment (VannStrasen) [reply src/skilltrees/skilltree.c]: Fixed in most recent commit.

**Comment by VannStrasen** [comment]
Good catch — I fixed those and found a couple more, so I edited those as well.

**Comment by firatciftci** [comment]
Issue Score: ✔️+

Comments:
Great work on this issue! The only thing I will point out is that the reviewer for this PR approved the PR without comment. Please make sure that you always do a full review of the code, even if the senior developer has already provided their review. In the rare cases where there truly isn't anything to improve in the PR, make sure to leave a comment confirming that you reviewed the code and don't have any changes to suggest. If the senior dev has left their review, it can also be useful to indicate whether you agree with it or not.

**Pull request merged**